### PR TITLE
chore : used the clear method

### PR DESCRIPTION
**Pull request by sammychinedu2ky**
@Rick-Anderson 
Fixes https://github.com/dotnet/AspNetCore.Docs/issues/28664

### CODE REVIEW

File: aspnetcore/performance/ObjectPool/ObjectPoolSample8/BirthdayMiddleware.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+                    var html = encoder.Encode(stringBuilder.ToString());
+                    await context.Response.WriteAsync(html);
+                }
+            }
+            finally // Ensure this runs even if the main code throws.
+            {
+                // Clear the string builder and return it to the pool.
+                stringBuilder.Clear();
+
```
Comment (Rick-Anderson) [inline aspnetcore/performance/ObjectPool/ObjectPoolSample8/BirthdayMiddleware.cs:65]:
```suggestion
```

**Review by Rick-Anderson — CHANGES_REQUESTED** [review]
Isn't this supposed to use the [New IResettable interface in ObjectPool](https://devblogs.microsoft.com/dotnet/asp-net-core-updates-in-dotnet-8-preview-2/#new-iresettable-interface-in-objectpool) ?

```
public class ReusableBuffer : IResettable
{
    public byte[] Data { get; } = new byte[1024 * 1024]; // 1 MB

    public bool TryReset()
    {
        Array.Clear(Data);
        return true;
    }
}

var bufferPool = ObjectPool.Create<ReusableBuffer>();

var buffer = bufferPool.Get();
try
{
    await ProcessDataAsync(buffer.Data);
}
finally
{
    bufferPool.Return(buffer); // Data is automatically reset
}
```

**Comment by sammychinedu2ky** [comment]
> Isn't this supposed to use the [New IResettable interface in ObjectPool](https://devblogs.microsoft.com/dotnet/asp-net-core-updates-in-dotnet-8-preview-2/#new-iresettable-interface-in-objectpool) ?
> 
> ```
> public class ReusableBuffer : IResettable
> {
>     public byte[] Data { get; } = new byte[1024 * 1024]; // 1 MB
> 
>     public bool TryReset()
>     {
>         Array.Clear(Data);
>         return true;
>     }
> }
> 
> var bufferPool = ObjectPool.Create<ReusableBuffer>();
> 
> var buffer = bufferPool.Get();
> try
> {
>     await ProcessDataAsync(buffer.Data);
> }
> finally
> {
>     bufferPool.Return(buffer); // Data is automatically reset
> }
> ```

The IResettable interface isn't available

**Comment by Rick-Anderson** [comment]
@xiaoyuvax do you know what's available for `IResettable`? I'm going off of  the [New IResettable interface in ObjectPool](https://devblogs.microsoft.com/dotnet/asp-net-core-updates-in-dotnet-8-preview-2/#new-iresettable-interface-in-objectpool)

**Comment by Rick-Anderson** [comment]
@sammychinedu2ky  see [this](https://devblogs.microsoft.com/dotnet/asp-net-core-updates-in-dotnet-8-preview-2/#comment-17636)

**Comment by Rick-Anderson** [comment]
@sammychinedu2ky  it works for me
`NuGet\Install-Package Microsoft.Extensions.ObjectPool -Version 8.0.0-preview.2.23153.2`

```
using Microsoft.Extensions.ObjectPool;

public class ReusableBuffer : IResettable
{
    public byte[] Data { get; } = new byte[1024 * 1024]; // 1 MB

    public bool TryReset()
    {
        Array.Clear(Data);
        return true;
    }
}
```

**Comment by sammychinedu2ky** [comment]
done @Rick-Anderson .. I didn't have the latest preview so that caused the initial error 😅

**Comment by Rick-Anderson** [comment]
@sammychinedu2ky  see [this](https://devblogs.microsoft.com/dotnet/asp-net-core-updates-in-dotnet-8-preview-2/#comment-17636)  I thought it didn't work with string builder.

**Comment by sammychinedu2ky** [comment]
> @sammychinedu2ky see [this](https://devblogs.microsoft.com/dotnet/asp-net-core-updates-in-dotnet-8-preview-2/#comment-17636) I thought it didn't work with string builder.

Yh String builder can't implement it but you can wrap the `StringBuilder` in a class that implements `IResettable`

**Comment by sammychinedu2ky** [comment]
@Rick-Anderson

**Comment by xiaoyuvax** [comment]
> @xiaoyuvax do you know what's available for `IResettable`? I'm going off of the [New IResettable interface in ObjectPool](https://devblogs.microsoft.com/dotnet/asp-net-core-updates-in-dotnet-8-preview-2/#new-iresettable-interface-in-objectpool)

sorry, I am a little bit out of context of this.  I read that updates for .NET 8 Preview 2, where it emphasizes little about the performance influence of IResettable. that is what i care most. Any details on performance issue of IResttable?

**Comment by Rick-Anderson** [comment]
@xiaoyuvax please review this sample code. It's just a sample on how to use the new interface.

**Comment by xiaoyuvax** [comment]
> @xiaoyuvax please review this sample code. It's just a sample on how to use the new interface.

I reviewd ObjectPoolSample6 (and ObjectPoolSample8 is but a stub), looks like sort of demo of a normal usage of ObjectPool instantiated with StringBuilder, but doesn't demonstrate neither the usage nor what's available for the IResettable interface. 
and then?

**Comment by Rick-Anderson** [comment]
Thanks @xiaoyuvax 
@sammychinedu2ky  can you show how to use it? The API page will show what's available.

**Comment by sammychinedu2ky** [comment]
> Rick

Show where?

**Comment by Rick-Anderson** [comment]
Per  the [New IResettable interface in ObjectPool](https://devblogs.microsoft.com/dotnet/asp-net-core-updates-in-dotnet-8-preview-2/#new-iresettable-interface-in-objectpool):

Apps can use an object pool if the values are expensive to allocate or initialize.

So have a method, `GetExpensiveAllocation(string key)`
where `GetExpensiveAllocation(string key)` just returns the `keyValue` of `key`+ "-" + `uppercase(key) + _ lowercase(key)`
 `GetExpensiveAllocation(string key)` isn't shown in article but you can use it for testing.

Get a  `keyValue` and log it. Show how to use the `IResettable ` interface

@xiaoyuvax  does that sound reasonable. Please add your comments.

**Comment by sammychinedu2ky** [comment]
please what do you mean by  keyValue of key+ "-" + uppercase(key) + _ lowercase(key) ?

**Comment by Rick-Anderson** [comment]
```
string key = "MyKey";
string keyValue = GetExpensiveAllocation(key);

string GetExpensiveAllocation(string key)
{
    return key + "_" + key.ToUpper() + "_" + key.ToLower();
}

Console.WriteLine(keyValue);
```

**Comment by sammychinedu2ky** [comment]
> ```
> string key = "MyKey";
> string keyValue = GetExpensiveAllocation(key);
> 
> string GetExpensiveAllocation(string key)
> {
>     return key + "_" + key.ToUpper() + "_" + key.ToLower();
> }
> 
> Console.WriteLine(keyValue);
> ```

ooh yeah that's right🚀

**Comment by sammychinedu2ky** [comment]
> ```
> string key = "MyKey";
> string keyValue = GetExpensiveAllocation(key);
> 
> string GetExpensiveAllocation(string key)
> {
>     return key + "_" + key.ToUpper() + "_" + key.ToLower();
> }
> 
> Console.WriteLine(keyValue);
> ```

ooh yeah that's right🚀

**Comment by xiaoyuvax** [comment]
Obviously, it's not proper to use StringBuilder for demonstration, for StringBuilder can't implement IResettable at all.
One should use a general class for demonstration, as the example given in [ASP.NET Core updates in .NET 8 Preview 2](https://devblogs.microsoft.com/dotnet/asp-net-core-updates-in-dotnet-8-preview-2/#new-iresettable-interface-in-objectpool)

According to [IResettable source](https://github.com/dotnet/aspnetcore/blob/d4be4957c24c7c8b745ade4cbaf290ad9cad1ad2/src/ObjectPool/src/IResettable.cs#L7), this interface contains only a sole definition of the method TryReset(). So we must describe how this method should be implemented and how it would be used in certain scenario.

Here's a more proper demo, which demonstrates completely what we can do with TryReset(), flexibly, such as that u don't have to reset everything as you can spare certain fields for some purpose:

```
public class ReusableObject : IResettable
{
    public  int ReuseCounter { get; private set; } = 0 ;  // this field will not be reset as to record how many times this object has been reused.
    public byte[] Buffer { get; } = new byte[1024 * 1024]; // 1 MB
    public string Title ="Some title that may be changed by user";

    ///This Method will be called by ObjectPool when object is returned.
    public bool TryReset()
    {
        ReuseCounter++;
        Array.Clear(Buffer); 
        Title ="";
        return true;   
    }
}

var bufferPool = ObjectPool.Create<ReusableObject >();

var buffer = bufferPool.Get();
try
{
    await ProcessDataAsync(buffer.Data);
}
finally
{
    bufferPool.Return(buffer); // Data is automatically reset
}
```

@Rick-Anderson is this what you want?

**Comment by Rick-Anderson** [comment]
@sammychinedu2ky  can you use code similar to what @xiaoyuvax suggests?

**Comment by sammychinedu2ky** [comment]
The one I wrote is similar or should I use exactly what he wrote? @Rick-Anderson

**Comment by Rick-Anderson** [comment]
> The one I wrote is similar or should I use exactly what he wrote? @Rick-Anderson

Do you disagree with what @xiaoyuvax  wrote above?

**Comment by sammychinedu2ky** [comment]
No, I don't disagree, I just made a custom wrapper of a string builder in mine using the IResettable Interface @Rick-Anderson

**Comment by sammychinedu2ky** [comment]
but I could change it to use his code if that is preferable

**Comment by Rick-Anderson** [comment]
@geeknoid  see [this comment](https://github.com/dotnet/AspNetCore.Docs/pull/28764#issuecomment-1486294690https://github.com/dotnet/AspNetCore.Docs/pull/28764#issuecomment-1486294690) or is the sample in this PR OK?

**Comment by JamesNK** [comment]
StringBuilder already has a way to reset its contents. See sample here: https://learn.microsoft.com/en-us/aspnet/core/performance/objectpool?view=aspnetcore-7.0#objectpool-sample. `StringBuilderPooledObjectPolicy` ensures the SB is reset.

The sample I added to the blog post - https://devblogs.microsoft.com/dotnet/asp-net-core-updates-in-dotnet-8-preview-2/#new-iresettable-interface-in-objectpool - is good because there is no way to reset a byte array otherwise. Wrapping it in a class than implements `IResettable` and then pooling the wrapped class seems like it works well. I just wrote a sample for the blog post and haven't used it in a real app though.

**Comment by sammychinedu2ky** [comment]
@Rick-Anderson please could you check it out? thanks

**Comment by Rick-Anderson** [comment]
@JamesNK  quick look please. Looks good to me.

File: aspnetcore/performance/ObjectPool/ObjectPoolSample8/Program.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+var builder = WebApplication.CreateBuilder(args);
+
+builder.Services.TryAddSingleton<ObjectPoolProvider, DefaultObjectPoolProvider>();
+
+builder.Services.TryAddSingleton<ObjectPool<ReusableBuffer>>(serviceProvider =>
+{
+    var provider = serviceProvider.GetRequiredService<ObjectPoolProvider>();
+    var policy = new ReusableBufferPolicy();
+    return provider.Create(policy);
```
Comment (JamesNK) [inline aspnetcore/performance/ObjectPool/ObjectPoolSample8/Program.cs:13]:
```suggestion
    var policy = new DefaultPooledObjectPolicy<ReusableBuffer>();
    return provider.Create(policy);
```

File: aspnetcore/performance/ObjectPool/ObjectPoolSample8/Program.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    {
+        reusableBuffer = new ReusableBuffer();
+        return reusableBuffer;
+    }
+
+    public bool Return(ReusableBuffer obj)
+    {
+        return obj.TryReset();
+    }
+}
```
Comment (JamesNK) [inline aspnetcore/performance/ObjectPool/ObjectPoolSample8/Program.cs:73]:
```suggestion
```

**Review by JamesNK — COMMENTED** [review]
You should be able to use the default policy: https://github.com/dotnet/aspnetcore/blob/8968058c9e5fdfdd1242426a03dc80609997edab/src/ObjectPool/src/DefaultPooledObjectPolicy.cs

**Comment by sammychinedu2ky** [comment]
> You should be able to use the default policy: https://github.com/dotnet/aspnetcore/blob/8968058c9e5fdfdd1242426a03dc80609997edab/src/ObjectPool/src/DefaultPooledObjectPolicy.cs

Thanks @JamesNK . let me fix that 🚀

**Comment by sammychinedu2ky** [comment]
Done @Rick-Anderson

File: aspnetcore/performance/ObjectPool/ObjectPoolSample8/Program.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+// return the SHA256 hash of a word 
+// https://localhost:7214/SamsonAmaugo
+app.MapGet("/{name}", async (string name, ObjectPool<ReusableBuffer> bufferPool) =>
+{
+
+    var buffer = bufferPool.Get();
+    try
+    {
+        // Set the buffer data to the ASCII values of a word
+        for (int i = 0; i < name.Length; i++)
```
Comment (JamesNK) [inline aspnetcore/performance/ObjectPool/ObjectPoolSample8/Program.cs:29]:
```suggestion
        for (var i = 0; i < name.Length; i++)
```

File: aspnetcore/performance/ObjectPool/ObjectPoolSample8/Program.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        {
+            buffer.Data[i] = (byte)name[i];
+        }
+
+        using (var sha256 = SHA256.Create())
+        {
+            // Calculate the hash of the data
+            var hash = sha256.ComputeHash(buffer.Data);
+            return "Hash: " + BitConverter.ToString(hash).Replace("-", "");
```
Comment (JamesNK) [inline aspnetcore/performance/ObjectPool/ObjectPoolSample8/Program.cs:38]:
```suggestion
            return "Hash: " + Convert.ToHexString(hash);
```
There is a convert method that doesn't include dashes: https://learn.microsoft.com/en-us/dotnet/api/system.convert.tohexstring?view=net-7.0

**Review by JamesNK — APPROVED** [review]

File: aspnetcore/performance/ObjectPool/ObjectPoolSample8/Program.cs (hunk 4 above)
Comment (sammychinedu2ky) [reply aspnetcore/performance/ObjectPool/ObjectPoolSample8/Program.cs:29]: Hi @JamesNK. just for clarification purposes and learning. Please what was your reason behind switching to var

File: aspnetcore/performance/ObjectPool/ObjectPoolSample8/Program.cs [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        }
```
Comment (JamesNK) [inline aspnetcore/performance/ObjectPool/ObjectPoolSample8/Program.cs:39]:
```suggestion
        Span<byte> hash = stackalloc byte[32];
        SHA256.HashData(buffer.Data.Slice(name.Length), hash);
        return "Hash: " + Convert.ToHexString(hash);
```
1. Prefer static hash data method https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1850
2. Only want to hash the part of the array that's used
3. Can stackalloc destination for hash calculation
4. There is a convert method that doesn't include dashes: https://learn.microsoft.com/en-us/dotnet/api/system.convert.tohexstring?view=net-7.0

A bunch of clean up. I think that's it 😄

Comment (sammychinedu2ky) [reply aspnetcore/performance/ObjectPool/ObjectPoolSample8/Program.cs]: @JamesNK Slice doesn't seem to be available for the array byte data type

File: aspnetcore/performance/ObjectPool/ObjectPoolSample8/Program.cs [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    try
+    {
+        // Set the buffer data to the ASCII values of a word
+        for (var i = 0; i < name.Length; i++)
+        {
+            buffer.Data[i] = (byte)name[i];
+        }
+
+        Span<byte> hash = stackalloc byte[32];
+        SHA256.HashData(new Span<byte>(buffer.Data,0,name.Length), hash);
```
Comment (JamesNK) [inline aspnetcore/performance/ObjectPool/ObjectPoolSample8/Program.cs:35]:
```suggestion
        SHA256.HashData(buffer.AsSpan(0, name.Length), hash);
```

**Comment by sammychinedu2ky** [comment]
thanks @JamesNK 🚀

**Pull request merged**